tests/e2e/slashing.go

```diff
@@ -374,106 +375,371 @@ func (suite *CCVTestSuite) TestOnRecvSlashPacketErrors() {
 	suite.Require().Equal(uint64(1), (providerKeeper.GetThrottledPacketDataSize(ctx, consumerChainID)))
 }
 
-// TestHandleSlashPacketDistribution tests the slashing of an undelegation balance
-// by varying the slash packet VSC ID mapping to infraction heights
-// lesser, equal or greater than the undelegation entry creation height
-func (suite *CCVTestSuite) TestHandleSlashPacketDistribution() {
-	providerKeeper := suite.providerApp.GetProviderKeeper()
-	providerStakingKeeper := suite.providerApp.GetE2eStakingKeeper()
-	providerSlashingKeeper := suite.providerApp.GetE2eSlashingKeeper()
-
-	// choose a validator
-	tmValidator := suite.providerChain.Vals.Validators[0]
-	valAddr, err := sdk.ValAddressFromHex(tmValidator.Address.String())
-	suite.Require().NoError(err)
-
-	validator, found := providerStakingKeeper.GetValidator(suite.providerChain.GetContext(), valAddr)
-	suite.Require().True(found)
+// TestSlashUndelegation tests the slashing of an undelegation balance in various scenarios
+func (suite *CCVTestSuite) TestSlashUndelegation() {
+	valIndex := 0
+	bondAmt := sdk.NewInt(10000000)
+	halfBondAmt := bondAmt.Quo(sdk.NewInt(2))
+	slashFactor := suite.providerApp.GetE2eSlashingKeeper().SlashFractionDowntime(suite.providerCtx())
+	slashAmountDec := slashFactor.MulInt(halfBondAmt)
+	slashAmount := slashAmountDec.TruncateInt()
+
+	var powerBeforeDelegate int64
+	var powerBeforeUndelegate int64
+	var powerAfterUndelegate int64
+	var delegateConsumerHeight uint64
+	var undelegateConsumerHeight uint64
+
+	consumerUnbondingPeriod := suite.consumerApp.GetConsumerKeeper().GetUnbondingPeriod(suite.consumerCtx())
+	fmt.Printf("consumerUnbondingPeriod: %s\n", consumerUnbondingPeriod)
+	providerUnbondingPeriod := suite.providerApp.GetE2eStakingKeeper().UnbondingTime(suite.providerCtx())
+	fmt.Printf("providerUnbondingPeriod: %s\n", providerUnbondingPeriod)
 
-	// unbonding operations parameters
-	delAddr := suite.providerChain.SenderAccount.GetAddress()
-	bondAmt := sdk.NewInt(1000000)
-
-	// new delegator shares used
-	testShares := sdk.Dec{}
-
-	// setup the test with a delegation, a no-op and an undelegation
-	setupOperations := []struct {
-		fn func(suite *CCVTestSuite) error
+	testCases := []struct {
+		name                  string
+		slash                 func(consAddr sdk.ConsAddress)
```

Comment:
It doesn't seem like we need to define `slash` and `checkDelegatorBalance` callbacks for each test case, since it's only the function parameters that are different between cases. Something similar to this is more understandable and prob less code:
```suggestion
		incrementBy time.Duration
		infractionHeight int64
		expectedEndBalance sdk.Int
		// etc. - shows what func parameters are different between test cases, no need to vary func calls
```

Using callbacks to define test cases should be used sparingly imo, as it makes things less readable unless the callbacks are brief

Reply:
Thanks for the fix :)